### PR TITLE
Use official Prometheus client, deprecate old one and StatsD

### PR DESCRIPTION
**Pull request by sbrunner**
- Using the standard lib
- Opening a separate port
- Work with the stats_db with push gateway or not
- Include booth metric (from StatsD and old Prometheus)
- Include standard Prometheus metrics
- Support multiprocess
- Support custom collectors by using Redis to collect all of them

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -79,9 +79,9 @@ pull: ## Pull the Docker images
 	for image in `find -name "docker-compose*.yaml" | xargs grep --no-filename "image:" | awk '{print $$2}' | sort -u | grep -v $(DOCKER_BASE) | grep -v rancher`; do docker pull $$image; done
 
 .PHONY: run
-run: build_test_app ## Run the test application
+run: build_test_app build_docker ## Run the test application
 	# cp acceptance_tests/tests/docker-compose.override.sample.yaml acceptance_tests/tests/docker-compose.override.yaml
-	cd acceptance_tests/tests/; TEST_IP=172.17.0.1 docker-compose up
+	cd acceptance_tests/tests/; docker-compose up --detach
```
Comment (gberaudo) [inline Makefile:84]: Personnaly I prefer not to detach because it allows to use "ctrl-c" to kill it, which the usual expectation.

File: README.md [hunk 2]
```diff
@@ -479,6 +483,92 @@ add_provider(MemoryMapProvider('rss'))
...
+
+In your `Dockerfile` you should add:
+
+```dockerfile
+RUN mkdir -p /prometheus-metrics \
+    chmod a+rwx /prometheus-metrics
+ENV PROMETHEUS_MULTIPROC_DIR=/prometheus-metrics
+```
+
+### Add custom metric collector
```
Comment (gberaudo) [inline README.md:537]: What is a custom metric collector?

File: acceptance_tests/tests/tests/test_stats.py [hunk 3]
```diff
@@ -1,6 +1,9 @@
+import pytest
+
 from c2cwsgiutils.acceptance import retry
 
 
+@pytest.mark.skip(reason="Deprecated")
```
Comment (gberaudo) [inline acceptance_tests/tests/tests/test_stats.py:6]: Deprecated means it still works and tested, but not future proof.
If it is not tested it should be removed.

File: acceptance_tests/tests/tests/test_stats_db.py [hunk 4]
```diff
@@ -1,33 +1,102 @@
...
+                "run_test",
+                "c2cwsgiutils-stats-db",
+                "--db=postgresql://www-data:www-data@db:5432/test",
+                "--schema=public",
+                "--extra=select 'toto, 42",
+                timeout=60,
+            )
+
+
+def test_standalone(prometheus_stats_db_connection, composition):
```
Comment (gberaudo) [inline acceptance_tests/tests/tests/test_stats_db.py:95]: Can you add a comment about what this does?

File: c2cwsgiutils/acceptance/composition.py [hunk 5]
```diff
@@ -64,15 +64,28 @@ def __init__(self, request: Request, composition: str, coverage_paths: Optional[
     def dc(self, args: List[str], **kwargs: Any) -> str:
         return cast(
             str,
-            subprocess.check_output(  # nosec
+            subprocess.run(  # nosec
```
Comment (gberaudo) [inline c2cwsgiutils/acceptance/composition.py:67]: Why this change?

File: c2cwsgiutils/acceptance/composition.py [hunk 6]
```diff
@@ -64,15 +64,28 @@ def __init__(self, request: Request, composition: str, coverage_paths: Optional[
...
                 env=Composition._get_env(),
                 stderr=subprocess.STDOUT,
+                stdout=subprocess.PIPE,
+                check=True,
                 **kwargs,
-            ).decode(),
+            ).stdout.decode(),
         )
 
+    def dc_proc(self, args: List[str], **kwargs: Any) -> subprocess.CompletedProcess[str]:
```
Comment (gberaudo) [inline c2cwsgiutils/acceptance/composition.py:78]: What does _proc mean here?

File: c2cwsgiutils/acceptance/composition.py [hunk 7]
```diff
@@ -86,9 +99,7 @@ def stop_all(self) -> None:
             os.makedirs(target_dir, exist_ok=True)
             for path in self.coverage_paths:
                 try:
-                    subprocess.check_call(  # nosec
-                        ["docker", "cp", path, target_dir], stderr=subprocess.STDOUT
-                    )
+                    subprocess.run(["docker", "cp", path, target_dir], check=True, timeout=60)  # nosec
```
Comment (gberaudo) [inline c2cwsgiutils/acceptance/composition.py:102]: Why removing stderr redirection?
Why not simply adding timeout=60 to `check_call`?

File: c2cwsgiutils/health_check.py [hunk 8]
```diff
@@ -548,6 +536,11 @@ def _maybe_function(what: Any, request: pyramid.request.Request) -> Any:
     return what(request) if callable(what) else what
 
 
+@broadcast.decorator(expect_answers=False)
+def _set_success(check_name: str) -> None:
+    _HEALTH_CHECKS_FAILURE.labels(name=check_name).set(0)
```
Comment (gberaudo) [inline c2cwsgiutils/health_check.py:541]: What does this do?

File: Makefile (hunk 1 above)
Comment (sbrunner) [reply Makefile:84]: Personally, I can't work without it because when I do ctrl-c to do another action it kills it...
In my workflow, this monopolizes one terminal for nothing :-(

File: README.md [hunk 9]
```diff
@@ -479,6 +483,92 @@ add_provider(MemoryMapProvider('rss'))
...
+    del server
+
+    registry = CollectorRegistry()
+    registry.register(MyCollector())
+    prometheus_client.start(registry)
+```
+
+### Usage of metrics
+
+With c2cwsgiutils each instance (Pod) has its own metrics, so we need to aggregate them to have the metrics for the service you probably need to use `sum by (<fields>) (<metric>)` to get the metric (especially for counters).
```
[the remark this answers is not in the document]
Comment (sbrunner) [reply README.md:570]: I just add a link to the official documentation:
https://github.com/prometheus/client_python#custom-collectors

File: acceptance_tests/tests/tests/test_stats_db.py [hunk 10]
```diff
@@ -1,33 +1,102 @@
...
+    def __enter__(self):
+        self.thread.start()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.stop = True
+
+
+def test_no_extra(app2_connection, prometheus_test_connection, composition):
+    with _Prometheus(prometheus_test_connection):
```
[the remark this answers is not in the document]
Comment (sbrunner) [reply acceptance_tests/tests/tests/test_stats_db.py:41]: > Also, can't this be achieved with a simple loop on the main thread?
No, because the process is blocked and waiting to be called...

File: c2cwsgiutils/acceptance/composition.py (hunk 5 above)
Comment (sbrunner) [reply c2cwsgiutils/acceptance/composition.py:67]: Because of the first sentence in the doc: https://docs.python.org/3/library/subprocess.html#using-the-subprocess-module

File: c2cwsgiutils/acceptance/composition.py (hunk 6 above)
Comment (sbrunner) [reply c2cwsgiutils/acceptance/composition.py:78]: That we return the CompletedProcess...

File: c2cwsgiutils/acceptance/composition.py (hunk 7 above)
Comment (sbrunner) [reply c2cwsgiutils/acceptance/composition.py:102]: Because of the first sentence in the doc: https://docs.python.org/3/library/subprocess.html#using-the-subprocess-module
Because we get the unwanted stderr...
My dream is to completely remove this file, I just get issue and hidden things with that way to do...

File: c2cwsgiutils/health_check.py (hunk 8 above)
Comment (sbrunner) [reply c2cwsgiutils/health_check.py:541]: Set check in success in all process.
I just add a comment.

File: c2cwsgiutils/prometheus_client.py [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+    if os.environ.get("PROMETHEUS_PORT") is not None:
+        broadcast.includeme()
+
+        registry = prometheus_client.CollectorRegistry() if registry is None else registry
+        registry.register(MemoryMapCollector())
+        registry.register(prometheus_client.PLATFORM_COLLECTOR)
+        registry.register(MultiProcessCustomCollector())
+        prometheus_client.multiprocess.MultiProcessCollector(registry)  # type: ignore[no-untyped-call]
+        prometheus_client.start_http_server(int(os.environ["PROMETHEUS_PORT"]), registry=registry)
```
[the remark this answers is not in the document]
Comment (sbrunner) [reply c2cwsgiutils/prometheus_client.py:35]: I assume that the port is not exposed to the world (not necessary related to k8s) :-)
I just added a remark in the README...

File: c2cwsgiutils/prometheus_client.py [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+    """Initialize prometheus_client in pyramid context."""
+
+    broadcast.subscribe("c2cwsgiutils_prometheus_collector_gc", _broadcast_collector_gc)
+    broadcast.subscribe("c2cwsgiutils_prometheus_collector_process", _broadcast_collector_process)
+
+
+def cleanup() -> None:
+    """Cleanup the prometheus_client registry."""
+
+    redis_utils.cleanup()
```
[the remark this answers is not in the document]
Comment (sbrunner) [reply c2cwsgiutils/prometheus_client.py:48]: To use the custom provider yes, also it's not required.

File: c2cwsgiutils/scripts/stats_db.py [hunk 13]
```diff
@@ -86,12 +104,21 @@ def do_report(
...
-        if self.prometheus is not None:
-            self.prometheus.commit()
+        if self.prometheus_push:
+            push_to_gateway(self.args.prometheus_url, job="db_counts", registry=self.registry)
+        elif self.statsd is None:
+            port = int(os.environ.get("PROMETHEUS_PORT", "9090"))
+            app = make_wsgi_app(self.registry)
+            with make_server("", port, app) as httpd:
+                print(f"Waiting that Prometheus get the metrics served on port {port}...")
+                httpd.handle_request()
```
[the remark this answers is not in the document]
Comment (sbrunner) [reply c2cwsgiutils/scripts/stats_db.py:121]: It handles one request then continues, yes it's relay needed...

**Comment by sbrunner** [comment]
Thanks for your comments @gberaudo I do a first quick pass, I will do the rest later :-)

File: Makefile (hunk 1 above)
Comment (gberaudo) [reply Makefile:84]: Yes, it monopolizes one terminal but we can have a lot of them ;)
And you get easy access to the logs.

If you like to detach, why don't you do `make run &`? That is the standard way.

File: c2cwsgiutils/acceptance/composition.py (hunk 6 above)
Comment (gberaudo) [reply c2cwsgiutils/acceptance/composition.py:78]: Then maybe let's call it `dc_process`.

File: c2cwsgiutils/acceptance/composition.py (hunk 7 above)
Comment (gberaudo) [reply c2cwsgiutils/acceptance/composition.py:102]: But then, if something goes wrong the output is silently drop, isn't it?

Do you have an example or more insights about the issues and hidden things you get?

> My dream is to completely remove this file, I just get issue and hidden things with that way to do...
Do you mean the composition mechamism?

File: c2cwsgiutils/health_check.py [hunk 14]
```diff
@@ -283,20 +302,9 @@ def __call__(self, request: pyramid.request.Request) -> str:
...
-                                "manual",
-                                "health_check",
-                                "alembic",
-                                alembic_ini_path,
-                                binding.name(),
-                            ]
-                            tags = None
-                        with stats.timer_context(key, tags):
+                        with _DB_TIMER.labels(
+                            configuration=alembic_ini_path, connection=binding.name(), check="alembic"
```
[the remark this answers is not in the document]
Comment (gberaudo) [reply c2cwsgiutils/health_check.py:306]: Creation time:
```python
_DB_TIMER = prometheus_client.Summary(
    os.environ.get("PROMETHEUS_PREFIX", "c2cwsgiutils_") + "health_check_db",
  "The to do a database query",
    ["configuration", "connection", "check"],
    unit="seconds",
```

Usage time:
```python
_DB_TIMER.labels(
                            configuration=alembic_ini_path, connection=binding.name(), check="alembic"
).time():
```

=> the `configuration`, `connection`, `check` params are the ones declared at creation time.
What happen if there is a mismatch?
If I pass ccccccconfiguration=alembic_ini_path?
If some parameter is missing?

File: c2cwsgiutils/prometheus_client.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (gberaudo) [reply c2cwsgiutils/prometheus_client.py:35]: See https://github.com/prometheus/client_python/blob/be3aac3f4c1c66f8f1a7b6b7d097d5b426201a6e/prometheus_client/exposition.py#L163

This is listening on all ipv4 interfaces by default.

File: acceptance_tests/tests/tests/test_stats.py (hunk 3 above)
Comment (gberaudo) [reply acceptance_tests/tests/tests/test_stats.py:6]: And also the existing /metrics prometheus endpoint.

File: c2cwsgiutils/prometheus_client.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (sbrunner) [reply c2cwsgiutils/prometheus_client.py]: Because Prometheus client didn't support Custom provider in multiprocess mode because I need that, I added a way to pass throw Redis to support it.

File: c2cwsgiutils/acceptance/utils.py [hunk 15]
```diff
@@ -45,7 +45,7 @@ def retry_timeout(what: Callable[[], Any], timeout: float = DEFAULT_TIMEOUT, int
         timeout: the timeout to get a success
         interval: the interval between try
     """
-    timeout = time.monotonic() + timeout
+    timeout = time.perf_counter() + timeout
```
[the remark this answers is not in the document]
Comment (sbrunner) [reply c2cwsgiutils/acceptance/utils.py:48]: I try the following code:
```python
import time


start = time.perf_counter()
for i in range(10000000):
    time.perf_counter()
print(time.perf_counter() - start)


start = time.perf_counter()
for i in range(10000000):
    time.monotonic()
print(time.perf_counter() - start)
```
And I get:
```
1.0459789808373898
1.045895482879132
```
It's so close that we can suspect that's the implementation is the same on Linux...

[the remark this answers is not in the document]
Comment (sbrunner) [reply c2cwsgiutils/acceptance/utils.py:48]: And:
```
>>> import time
>>> time.get_clock_info('monotonic')
namespace(implementation='clock_gettime(CLOCK_MONOTONIC)', monotonic=True, adjustable=False, resolution=1e-09)
>>> time.get_clock_info('perf_counter')
namespace(implementation='clock_gettime(CLOCK_MONOTONIC)', monotonic=True, adjustable=False, resolution=1e-09)
```
That probably means that the implementation is relay the same.

Then this change didn't do anything, I just prefer to be like that to be coherent with prometheus_client :-)

[the remark this answers is not in the document]
Comment (gberaudo) [reply c2cwsgiutils/acceptance/utils.py:48]: ok

**Review by gberaudo — APPROVED** [review]

**Comment by sbrunner** [comment]
Thanks @gberaudo
